### ISSUE TITLE
Issue with finding values with encrypted data

### ISSUE THREAD

**Issue by Amraneze** [abec/typeorm-encrypted#47]
I have encrypted a column `email` but I can't find user that have the same email because it gives a different encryption. Is there a way to bypass that ?

I tried something like this 

![image](https://user-images.githubusercontent.com/28459763/148827523-fb59735c-e0e8-4989-ac10-d7faeff11e33.png)

![image](https://user-images.githubusercontent.com/28459763/148827552-1590ba07-f97f-4e6a-bf21-ec6c0e8a8eeb.png)

But without a chance.

**Comment by generalpiston** [comment]
Take a peek at this test case: https://github.com/generalpiston/typeorm-encrypted/blob/master/test/find-operator.test.ts#L41

**Comment by Amraneze** [comment]
I used `Equal` but I'm getting the same result.

![image](https://user-images.githubusercontent.com/28459763/148987828-e392ea4f-6bcd-408a-a3c6-dfbb973842df.png)

**Comment by generalpiston** [comment]
@Amraneze So, the way it works is it encrypts the field and checks it against the internal value. If the secrets / keys have changed, then it won't work. Unfortunately, without access to your data set, I can't really figure it out for you. The tests we have for the find operator are passing. So, I suspect your key changed or the data doesn't exist.

**Comment by Amraneze** [comment]
@generalpiston if I use a static IV then it works fine. But, it's not a good practice to use a static one. The secret is set in environment variable.

![image](https://user-images.githubusercontent.com/28459763/149009025-ed50ca52-362f-4e0b-be51-4d8bb72bf05f.png)

**Comment by generalpiston** [comment]
@Amraneze hey that's a great point. Perhaps the IV should be different for each record?

**Comment by Amraneze** [comment]
@generalpiston I will just use a different IV for each entity. I can create a test project if you want to reproduce the issue.

>PS: I also enabled encryption on the database.

**Comment by generalpiston** [comment]
@Amraneze I don't think we need a test project. Every thing you said tracks. I'm going to think about the security implications of using the same IV a bit and get back to this.

**Comment by razr22** [comment]
Any update on this? I'm not very keen on using the same IV or IV's for each entity.

**Comment by Amraneze** [comment]
@razr22 I'm using the same IV. it's better to generate IV for each entity but it you will have a lot of IV in your env variables. @generalpiston should we close this issue ?

**Comment by generalpiston** [comment]
@Amraneze @razr22 having a different IV is interesting since it would reduce the cardinality of sets open to attack. I think to do this, we'd have to do one of the following:

1. Use another table.
2. Use another column in the same table.
3. Prefix the IV to encrypted value.

In some cases, the IV may be a counter. In others, it should be a uniformly randomly generated value. This would vary based on the encryption strategy.

This is kind of a big task... I won't have time to get to this for a while. @Amraneze or @razr22 would either of you feel comfortable implementing it?

Here's how I'd do it:

- Implement 3 from the above list for a very specific, commonly used encryption scheme (ie. AES CBC).
- Fallback to common IV.

I'm happy to add whoever implements this to the list of collaborators and create a list of contributors and list their name.

**Comment by coderdan** [comment]
I've just started using this lib and came across this issue. Please do not use a fixed IV with AES. While there are some very specific cases that it will be OK, it will almost always completely break the encryption. See https://crypto.stackexchange.com/questions/26790/how-bad-it-is-using-the-same-iv-twice-with-aes-gcm

Queryable encryption is generally not possible with correct, randomised encryption. You can use the encryption in deterministic mode which is more secure but if that is the case, the IV should be generated using something like the HMAC of the value (not fixed).

**Comment by generalpiston** [comment]
@coderdan thanks for the insight. Are you recommending we use randomly generated IVs?

**Comment by mpalmer** [comment]
Yes, you should absolutely use IVs generated from a cryptographically-secure random number generator for each record, and store them with the ciphertext.

**Comment by coderdan** [comment]
> @coderdan thanks for the insight. Are you recommending we use randomly generated IVs?

IVs must _never_ repeat. If the same message is encrypted more than once with the same IV you can literally decrypt the message by XOR'ing the ciphertexts.

A cryptographically strong randomly generated IV is the best solution. A counter can also be used but you need to make sure that it never resets (or wraps around) and can't be manipulated so much less preferred IMHO. Users of this library likely won't know the nuances of the issue so best to abstract IV management away and just generate the IV randomly internally.

That actually happens now if you don't pass an IV to the transformer so a simple fix here could just be to remove the ability to manually set an IV.

**Comment by generalpiston** [comment]
@coderdan I guess with GCM mode, the nonce can't repeat. In CBC mode, a repeat isn't catastrophic, but it's slower. We should probably allow the `iv` function to be an async function.

**Comment by coderdan** [comment]
With CBC it is _less_ bad, yes. Still a fixed IV will result in deterministic encryption and potential chosen plaintext attacks. Allowing the IV to be an async function would be OK so long as the risks are well documented :)

As you can see I get very passionate about this!

**Comment by bilalanees98** [comment]
@generalpiston is there a solution for this? sort of stuck on it.
